### ISSUE TITLE
Tilde `~` sometimes not quoted

### ISSUE THREAD

**Issue by Ordoviz** [ljharb/shell-quote#9]
```js
import {quote} from "shell-quote";
console.log(quote(['cat', '~/.bashrc']));   // tilde not quoted
console.log(quote(['cat', '~/s p a c e'])); // tilde quoted
```

The treatment of `~` should be consistent at the very least. I find it surprising that `~` is not quoted since all other special Bash characters are quoted. 

I suggest adding `~` to the list of characters to quote in this line:
 https://github.com/ljharb/shell-quote/blob/da8a3abb4b2754dd9c1d3203142133e253c515a4/quote.js#L14
 
 This quirk was showcased last week in [a AmateursCTF challenge](https://github.com/D13David/ctf-writeups/blob/main/amateursctf23/web/uwuctf/README.md).

**Comment by ljharb** [comment]
`~` isn't universally replaced with `$HOME`, i believe. Which shells besides bash support that?

**Comment by Ordoviz** [comment]
>Which shells besides bash support that?

Zsh, Fish, dash, oilshell, nushell, PowerShell and probably more

**Comment by ljharb** [comment]
node, however, does not, i think, if you pass it in child_process.exec?

**Comment by Ordoviz** [comment]
> node, however, does not, i think, if you pass it in child_process.exec?

```js
const { exec } = require('node:child_process');
exec('echo ~', (error, stdout, stderr) => {
  console.log(`exec: ${stdout}`);
}); 

const { execFile } = require('node:child_process');
execFile('echo', ['~'], {shell: false}, (error, stdout, stderr) => {
  console.log(`execFile (shell=false): ${stdout}`);
}); 
execFile('echo', ['~'], {shell: true}, (error, stdout, stderr) => {
  console.log(`execFile (shell=true): ${stdout}`);
}); 
```

Unsurprisingly, the output is
```
execFile (shell=true): /home/me

execFile (shell=false): ~

exec: /home/me
```

**Comment by ljharb** [comment]
ahh, the `shell` option is the difference :-) i've never passed that, so i've always been confused that `~` doesn't expand.

sounds like this would be a useful thing to quote. If we did it automatically, in what ways could that be a breaking change? (the case above where it quotes is because of the spaces, so it's not inconsistent right now)

**Comment by Ordoviz** [comment]
>in what ways could that be a breaking change? 

Looking at the code, we see that `~` is already quoted when the input contains whitespace, a single quote, or a double quote. Otherwise we reach the `return String(s).replace` line, which does not quote `~`.

https://github.com/ljharb/shell-quote/blob/da8a3abb4b2754dd9c1d3203142133e253c515a4/quote.js#L11-L14

Speaking of introducing breaking changes, you could consider quoting the percent sign `%` as well for good measure, as mentioned on [Exploiting CVE-2021-42740](https://wh0.github.io/2021/10/28/shell-quote-rce-exploiting.html#notes-on-exploiting).

**Comment by ljharb** [comment]
Right, but the presence of `~` isn't triggering the quoting, which remains consistent currently - otherwise, you could claim that `a` is quoted on line 12 but not on line 14, which is true but a useless thing to measure.

The goal is to *not* introduce breaking changes, since that's the most harmful thing any package can do to the ecosystem.

My question was, if we change it so that `~` (or `%`) alone can trigger the quoting, what would that break for a user?

**Comment by ljharb** [comment]
This is fixed as of v1.9.0: `~` is now escaped, so it can't trigger tilde expansion regardless of what else is in the string.

Your original cases:

```
quote(['cat', '~/.bashrc'])   => cat \~/.bashrc
quote(['cat', '~/s p a c e']) => cat '~/s p a c e'
```

Both now keep the tilde literal, so the inconsistency (and the CTF-style expansion) is gone. Thanks for the detailed thread, and for confirming the `shell: true` behavior in `child_process`; that's what made it clear this was worth escaping.

On the `%` suggestion, I'm going to leave that one alone. Unlike `~` (and the `^`/`@` we escape for zsh globbing), `%` is not a metacharacter in any POSIX-family shell as a command argument; it stays literal in bash, sh, and zsh:

```
$ set -- 100% %PATH% a%b; printf '[%s]' "$@"
[100%][%PATH%][a%b]
```

`%` is only special in `cmd.exe` (`%VAR%`), and cmd.exe is out of scope for this library, which emits POSIX quoting (see the README note). Escaping `%` would just sprinkle backslashes into percentages and encoded URLs for no POSIX benefit. And the CVE-2021-42740 reference was a `parse` issue (fixed back in 1.7.3), not a `quote` escaping gap.

Closing as fixed. Thanks again!
